CRM/Event/BAO/Participant.php

```diff
@@ -832,6 +844,324 @@ public static function deleteParticipant($id) {
+   * @param array $excludeStatus
+   *   A list of registration status to not consider (e.g. for ignoring cancelled registrations).
+   * @param array $filterRoleIds
+   *   A list of participant role IDs to filter for. Registrations with other roles will not be considered.
+   * @param bool $includeTest
+   *   Whether to include test participants.
+   *
+   * @return array<int,array{id:int,"status_id:name":string}>
+   *   An array of participants (a subset of attributes) matching the given criteria, keyed by ID.
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\UnauthorizedException
+   *
+   * @internal
+   */
+  public static function getExistingParticipants(
+    int $contactId,
+    int $eventId,
+    bool $onlyCounted = TRUE,
+    bool $includeOnWaitlist = TRUE,
+    array $excludeStatus = ['Cancelled'],
+    array $filterRoleIds = [],
+    bool $includeTest = FALSE
+  ) {
+    $query = \Civi\Api4\Participant::get(FALSE)
+      ->addSelect('id', 'status_id:name')
+      ->addWhere('contact_id', '=', $contactId)
+      ->addWhere('event_id', '=', $eventId);
+
+    if ($onlyCounted) {
+      $query
+        ->addJoin('ParticipantStatusType AS participant_status_type', 'LEFT');
+      $clauses = [
+        ['participant_status_type.is_counted', '=', TRUE],
+      ];
+      if ($includeOnWaitlist) {
+        $clauses = ['participant_status_type.name', '=', 'On waitlist'];
+      }
+      $query->addClause('OR', $clauses);
+    }
+
+    if ([] !== $excludeStatus) {
+      $query->addWhere('status_id:name', 'NOT IN', $excludeStatus);
+    }
+
+    if ([] !== $filterRoleIds) {
+      $query->addWhere('role_id', 'IN', $filterRoleIds);
+    }
+
+    if (!$includeTest) {
+      $query->addWhere('is_test', '=', FALSE);
+    }
+
+    $result = $query->execute();
+    return $result->getArrayCopy();
+  }
+
+  /**
+   * Checks whether a participant for a given event and contact exists.
+   * Can be used during validation of new event registrations to check for duplicates.
+   *
+   * @param int $contactId
+   *    The contact ID of participants to find.
+   * @param int $eventId
+   *    The event ID of participants to find.
+   * @param bool $onlyCounted
+   *    Whether to only consider registrations with a status with "is_counted".
+   * @param bool $includeOnWaitlist
+   *    Whether to consider registrations with status "On waitlist" when restricing to "is_counted".
+   * @param array $excludeStatus
+   *    A list of registration status to not consider (e.g. for ignoring cancelled registrations).
+   * @param array $filterRoleIds
+   *    A list of participant role IDs to filter for. Registrations with other roles will not be considered.
+   * @param bool $includeTest
+   *    Whether to include test participants.
+   *
+   * @return bool
+   * @throws \CRM_Core_Exception
+   * @throws \Civi\API\Exception\UnauthorizedException
+   *
+   * @internal
+   */
+  public static function exists(
```

Comment:
As so often happens in a refactor the function that you start with extracting seems less useful by the end - this function is only called from one place now - so I would suggest folding it back into that place